File: examples/py/websocket-orderbook-subscription.py
import os
import sys
import pprint
import traceback

pp = pprint.PrettyPrinter(depth=6)

root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(root + '/python')
# import ccxt  # noqa: E402
import ccxt.async_support as ccxt  # noqa: E402
import asyncio  # noqa: E402

loop = asyncio.get_event_loop()
# import txaio
# txaio.start_logging(level='debug')


async def main():
    if len(sys.argv) <= 5:
        print('python ' + __file__ + ' exchange apikey secret limit symbol ...')
        sys.exit(-1)

    exchange_id = sys.argv[1]
    apiKey = sys.argv[2]
    secret = sys.argv[3]
    limit = int(sys.argv[4])
    symbols = []
    for i in range(5, len(sys.argv)):
        symbols.append(sys.argv[i])

    exchange = getattr(ccxt, exchange_id)({
        "apiKey": apiKey,
        "secret": secret,
        "enableRateLimit": True,
        'verbose': True,
        'timeout': 5 * 1000,
        # 'wsproxy': 'http://185.93.3.123:8080/',
    })

    @exchange.on('err')
    def websocket_error(err, conxid):  # pylint: disable=W0612
        print(type(err).__name__ + ":" + str(err))
        traceback.print_tb(err.__traceback__)
        traceback.print_stack()
        loop.stop()

    @exchange.on('ob')
    def websocket_ob(symbol, ob):  # pylint: disable=W0612
        print("ob received from: " + symbol)
        sys.stdout.flush()
        # pp.pprint(ob)

    sys.stdout.flush()

    for j in range(2):
        for i in range(len(symbols)):
            symbol = symbols[i]
            print("subscribe: " + symbol)
            sys.stdout.flush()
            await exchange.websocket_subscribe('ob', symbol, {'limit': limit})
            print("subscribed: " + symbol)
            sys.stdout.flush()
            ob = await exchange.websocket_fetch_order_book(symbol, limit)  # noqa: F841 pylint: disable=W0612
            print("ob fetched: " + symbol)
            # print(ob)
            sys.stdout.flush()
            await asyncio.sleep(5)

        for i in range(len(symbols)):
            symbol = symbols[i]
            print("unsubscribe: " + symbol)
            sys.stdout.flush()
            await exchange.websocket_unsubscribe('ob', symbol)
            print("unsubscribed: " + symbol)
            sys.stdout.flush()
            await asyncio.sleep(5)

    await exchange.close()

loop.run_until_complete(main())
# loop.run_forever()
# loop.stop()
# loop.close()
print("after complete")

File: examples/js/websocket-orderbook-subscription.js
'use strict';

const asTable = require('as-table'),
  log = require('ololog').noLocate,
  ansi = require('ansicolor').nice,
  ccxt = require('../../ccxt.js'),
  cex = require('../../js/cex');
//const HttpsProxyAgent = require('https-proxy-agent');

function printUsage() {
  log(
    'Usage: node',
    process.argv[1],
    'exchange',
    'apiKey',
    'secret',
    'limit',
    'symbol',
    '...'
  );
}
let sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

let exchange;
async function fetchOrderBook(id, apiKey, secret, limit, symbols, params) {
  // const proxy = process.env.http_proxy || 'http://localhost:7080'; // HTTP/HTTPS proxy to connect to（only local http proxy work at "here"）
  // const agent = new HttpsProxyAgent(proxy);
  exchange = new ccxt[id]({
    apiKey: apiKey,
    secret: secret,
    enableRateLimit: true,
    verbose: true,
    // agent: agent
  });
  exchange.on('err', (err, conxid) => {
    try {
      console.log(err);
      exchange.websocketClose(conxid);
    } catch (ex) {
      console.log(ex);
    }
  });
  exchange.on('ob', (market, ob) => {
    console.log('ob received: ', market, ob);
    // console.log (ob);
  });
  await exchange.loadMarkets();

  for (let j = 0; j < 2; j++) {
    for (let i = 0; i < symbols.length; i++) {
      let symbol = symbols[i];
      console.log('subscribe: ' + symbol);
      await exchange.websocketSubscribe('ob', symbol, params);
      console.log('subscribed: ' + symbol);
      let ob = await exchange.websocketFetchOrderBook(symbol, limit);
      console.log('ob fetched: ' + symbol, ob);
      // console.log (ob);
      await sleep(5 * 1000);
    }

    for (let i = 0; i < symbols.length; i++) {
      let symbol = symbols[i];
      console.log('unsubscribe: ' + symbol);
      await exchange.websocketUnsubscribe('ob', symbol, params);
      console.log('unsubscribed: ' + symbol);
      await sleep(5 * 1000);
    }
  }
}

(async function main() {
  try {
    if (process.argv.length > 6) {
      const id = process.argv[2]
      const apiKey = process.argv[3]
      const secret = process.argv[4]
      const limit = parseInt (process.argv[5])
      const symbols = [];
      for (let i = 6 ; i < process.argv.length; i++) {
          symbols.push (process.argv[i].toUpperCase ())
      }
      //
      // this call gets live orderbook 
      const ob = await fetchOrderBook(id, apiKey, secret, limit, symbols, {
        // contract_type: 'next_week',
        'limit': limit,
      });
    } else {
      printUsage ()
    }
  } catch (ex) {
    log('Error:'.red, ex);
    log(ex.stack);
    exchange.websocketClose();
  }
  // process.exit ()
})();
